### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by SebastianSpeitel**
PR tracking refactor changes.

Steps -
 - [x] Move fns outside of utils
   - [x] Duration
   - [x] PlainDate
   - [x] Instant
   - [x] PlainDateTime
   - [x] PlainMonthDay
   - [x] PlainDateTime
- [x] Post util movement tidyup
- [x] Separate utils from date-specific utils
- [x] Perform 'in file' refactoring, e.g. within PlainDate can the internal helper functions (e.g. differenceDate) be folder into the implementation

### CODE REVIEW

**Comment by SebastianSpeitel** [comment]
Should we move all the "string-stuff" (parsing and formatting) out of utils?

File: assembly/plainyearmonth.ts [hunk 1]
```diff
@@ -265,13 +267,21 @@ export class PlainYearMonth {
...
+    return compare(
+      [a.year,
       a.month,
-      a.referenceISODay,
-      b.year,
+      a.referenceISODay],
+      [b.year,
       b.month,
-      b.referenceISODay
+      b.referenceISODay]
```
Comment (MaxGraey) [inline assembly/plainyearmonth.ts:276]: Using arrays here is unnecessary. I will only create additional gc allocs and slowdowns

File: assembly/util/index.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+// @ts-ignore: decorator
+@inline
+export function coalesce<T extends number>(a: T, b: T, nill: T = -1 as T):T {
+  return a == nill ? b : a;
+}
+
+// returns the larger of two values for ordered enums
+export function larger<T>(t1: T, t2: T): T {
+  return min(t1, t2) as T
+}
```
Comment (MaxGraey) [inline assembly/util/index.ts:87]: Why larger if `min` return smallest one of two argument? And why this util necessary?

Comment (SebastianSpeitel) [reply assembly/util/index.ts:87]: This was refactored from a function to order enums, where the enum values were ordered smallest to largest.
I don't like it, too.
Either just inline the `min` call and comment it or make a specific function for that special enum.

**Pull request merged**